File: kcpd_data.py
# Databricks notebook source
import csv
import pyspark.pandas as ps
from datetime import datetime
from sodapy import Socrata

# COMMAND ----------

client = Socrata("data.kcmo.org", None)
results = client.get_all("isbe-v4d8")
results_df = ps.DataFrame.from_records(results)

# COMMAND ----------

file_name = "isbe-v4d8" + datetime.today().strftime("%Y-%m-%d")
results_df.to_parquet(
    "/Volumes/datakc/volumes/isbe-v4d8/" + file_name, compression="snappy"
)

# COMMAND ----------

#TODO finish getting lat/long FOR EACH row
results_df['latitude'] = results_df[["location"]].iloc[1].iloc[0]["coordinates"][1]
results_df['longitude'] = results_df[["location"]].iloc[1].iloc[0]["coordinates"][0]
results_df = results_df.drop(['location',':@computed_region_kk66_ngf4',':@computed_region_k7in_q28t',':@computed_region_w4hf_t6bp',':@computed_region_qizh_zmq5',':@computed_region_vrdq_ghvi'], axis=1)

# COMMAND ----------

File: kcpd_data.r
# Databricks notebook source
suppressMessages(library(tidyverse))
suppressMessages(library(sparklyr))
suppressMessages(library(RSocrata))
suppressMessages(library(DBI))
suppressMessages(library(glue))
suppressMessages(library(arrow))

# COMMAND ----------

# Connect sparklyr to your spark cluster
sc <- spark_connect(method = "databricks")

# COMMAND ----------

# Read CSV from API and manipulate before saving
message("Starting data read")
results_df <- read.socrata("https://data.kcmo.org/resource/isbe-v4d8.json")
message("Data read completed")

# COMMAND ----------

# Transforming coordinate list to individual long and lat columns and binding back to original table (dropping nulls ~ 2.2%)
results_df <- results_df %>%
select(location.coordinates) %>%
filter(location.coordinates != "NULL") %>%
mutate(location.coordinates = map(location.coordinates, set_names, c("long","lat"))) %>%
unnest_wider(col = c("location.coordinates")) %>%
cbind(results_df %>% filter(location.coordinates != "NULL") %>% select(!c(location.coordinates,location.type)))

# COMMAND ----------

# Read CSV from API and write parquet to volume
file_name = glue("/Volumes/datakc/volumes/isbe-v4d8/isbe-v4d8_{Sys.Date()}")
message("Writing parquet file")
write_parquet(results_df, file_name)
message("Parquet file written")

# COMMAND ----------

# Reading parquet to spark and cleaning data
isbev4d8_spark <- spark_read_parquet(sc, file_name)
isbev4d8_spark <- isbev4d8_spark %>%
filter(!is.na(to_date)) %>%
mutate(duration = to_date - from_date)

# COMMAND ----------

# Creating table
isbev4d8_spark %>% spark_write_table('datakc.tables.kcpd_crime_data', mode = 'overwrite')

# COMMAND ----------


